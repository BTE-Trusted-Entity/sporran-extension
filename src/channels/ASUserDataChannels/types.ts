import { ICredential, KiltAddress } from '@kiltprotocol/types';
import { HexString } from '@polkadot/util/types';

import { DAppName } from '../AccessChannels/DAppName';
import { Origin } from '../AccessChannels/Origin';

export type ASUserDataInput = DAppName & {
  submitter: KiltAddress;
};

export type ASUserDataOriginInput = ASUserDataInput & Origin;

export interface ASUserDataOutput {
  createDidExtrinsic: HexString;
  credential: ICredential;
}
