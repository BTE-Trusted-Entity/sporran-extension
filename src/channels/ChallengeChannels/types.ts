import { DidResourceUri } from '@kiltprotocol/sdk-js';

export interface ChallengeInput {
  dAppEncryptionKeyId: DidResourceUri;
  challenge: string;
}

export interface ChallengeOutput {
  encryptionKeyId: DidResourceUri;
  encryptedChallenge: string;
  nonce: string;
}
