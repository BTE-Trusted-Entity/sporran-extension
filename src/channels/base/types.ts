export interface ErrorFirstCallback<Output> {
  (error: null, output: Output): void;

  (error: Error): void;
}
