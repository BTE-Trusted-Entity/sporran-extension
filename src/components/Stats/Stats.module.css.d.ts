export const __esModule: true;
export const stats: string;
export const balance: string;
