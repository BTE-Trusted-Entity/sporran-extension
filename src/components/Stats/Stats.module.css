.stats {
  composes: smallText from '../Typography/Typography.module.css';
  position: absolute;
  bottom: 0;
  align-self: center;
}

.balance {
  color: var(--color-lightest);
}
