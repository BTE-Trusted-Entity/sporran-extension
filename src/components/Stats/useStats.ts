import { useEffect, useRef, useState } from 'react';
import BN from 'bn.js';
import { find, mapValues, omit } from 'lodash-es';

import { IdentitiesMap } from '../../utilities/identities/identities';
import {
  BalanceChange,
  onAddressBalanceChange,
} from '../../utilities/balanceChanges/balanceChanges';

interface Balances {
  [address: string]: BN | null;
}

interface Subscriptions {
  [address: string]: () => void;
}

function subscribeToBalance(
  address: string,
  setBalances: (callback: (balances: Balances) => Balances) => void,
  subscriptions: Subscriptions,
) {
  async function handleBalance(error: Error | null, balance?: BalanceChange) {
    if (error || !balance) {
      console.error(error);
      return;
    }
    if (balance.address !== address) {
      return;
    }
    setBalances((balances: Balances) => ({
      ...balances,
      ...(address in balances && { [address]: balance.balances.total }),
    }));
  }

  const unsubscribe = onAddressBalanceChange(address, handleBalance);

  subscriptions[address] = () => {
    unsubscribe();
    delete subscriptions[address];
    setBalances((balances: Balances) => omit(balances, address));
  };
}

export function useStats(
  identities: IdentitiesMap,
): { count: number; total: BN } | null {
  const identitiesList = Object.values(identities);

  const [balances, setBalances] = useState<Balances>(
    mapValues(identities, () => null),
  );

  // A more straightforward way would be to use state, but that would cause unsubscribing everything
  // on every identity list change.
  const subscriptions = useRef<Subscriptions>({}).current;
  useEffect(() => {
    return () => {
      Object.values(subscriptions).forEach((unsubscribe) => unsubscribe());
    };
  }, [subscriptions]);

  // On every identity list change identify which addresses have been added or removed,
  // and subscribe or unsubscribe accordingly.
  useEffect(() => {
    identitiesList.forEach(({ address }) => {
      const addressWasAdded = !(address in subscriptions);
      if (addressWasAdded) {
        subscribeToBalance(address, setBalances, subscriptions);
      }
    });

    Object.keys(subscriptions).forEach((address) => {
      const addressWasRemoved = !find(identitiesList, { address });
      if (addressWasRemoved) {
        subscriptions[address]();
      }
    });
  }, [identitiesList, subscriptions]);

  const balancesList = Object.values(balances);
  if (balancesList.includes(null)) {
    return null; // do not show the total until all requests are fulfilled
  }

  const count = identitiesList.length;
  const total = new BN(0);
  balancesList.forEach((balance) => balance && total.iadd(balance));

  return { count, total };
}
