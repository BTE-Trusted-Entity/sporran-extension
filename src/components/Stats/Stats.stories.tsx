import { Meta } from '@storybook/react';

import { Stats } from './Stats';

export default {
  title: 'Components/Stats',
  component: Stats,
} as Meta;

export function Template(): JSX.Element {
  return <Stats />;
}
