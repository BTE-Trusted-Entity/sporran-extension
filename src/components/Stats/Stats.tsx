import { browser } from 'webextension-polyfill-ts';

import * as styles from './Stats.module.css';

import { plural } from '../../utilities/plural/plural';
import { KiltAmount } from '../KiltAmount/KiltAmount';
import {
  IdentitiesMap,
  useIdentities,
} from '../../utilities/identities/identities';

import { useStats } from './useStats';

interface Props {
  identities: IdentitiesMap;
}

function UnconditionalStats({ identities }: Props): JSX.Element | null {
  const t = browser.i18n.getMessage;

  const stats = useStats(identities);
  if (!stats) {
    return null; // hide the component while data is pending
  }

  return (
    <p className={styles.stats}>
      {plural(stats.count, {
        one: 'component_Stats_identity_one',
        other: 'component_Stats_identity_other',
      })}
      <span className={styles.balance}>
        {t('component_Stats_balance')}
        <KiltAmount amount={stats.total} type="funds" />
      </span>
    </p>
  );
}

export function Stats(): JSX.Element | null {
  const identities = useIdentities().data;
  if (!identities) {
    return null; // storage data pending
  }

  if (Object.values(identities).length < 2) {
    return null; // hide the component when too few identities
  }

  return <UnconditionalStats identities={identities} />;
}
