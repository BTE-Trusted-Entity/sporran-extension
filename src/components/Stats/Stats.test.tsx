import { IdentitiesProviderMock, render, screen } from '../../testing/testing';
import { IdentitiesMap } from '../../utilities/identities/types';

import { Stats } from './Stats';

describe('Stats', () => {
  it('should not render for no identities', async () => {
    const { container } = render(
      <IdentitiesProviderMock identities={{}}>
        <Stats />
      </IdentitiesProviderMock>,
    );
    expect(container).toBeEmptyDOMElement();
  });

  it('should not render for single identity', async () => {
    const identities: IdentitiesMap = {
      '4tDjyLy2gESkLzvaLnpbn7N61VgnwAhqnTHsPPFAwaZjGwP1': {
        name: 'Light DID Identity',
        address: '4tDjyLy2gESkLzvaLnpbn7N61VgnwAhqnTHsPPFAwaZjGwP1',
        did: 'did:kilt:light:004oUiK3EvTczc6ukYLEV57BuxzkW9zksKRchZXYGSocLmWE5N',
        index: 1,
      },
    };
    const { container } = render(
      <IdentitiesProviderMock identities={identities}>
        <Stats />
      </IdentitiesProviderMock>,
    );
    expect(container).toBeEmptyDOMElement();
  });

  it('should render for at least two identities', async () => {
    const { container } = render(<Stats />);
    expect(await screen.findByText(/Balance/)).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });
});
