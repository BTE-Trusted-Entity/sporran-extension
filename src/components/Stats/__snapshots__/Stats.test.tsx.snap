// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Stats should render for at least two identities 1`] = `
<div>
  <p
    class="stats"
  >
    You have 3 Identities with a
    <span
      class="balance"
    >
       Total Balance: 
      <span
        class="amount"
      >
        3.7020
         
        <span
          aria-label="KILT Coin"
          class="coin"
        />
      </span>
    </span>
  </p>
</div>
`;
